### PR TITLE
fixes #2389

### LINKED ISSUE #2389: [BUG] Bloated MoE expert checkpoint files with ZeRO 2
**Describe the bug**
It appears that with ZeRO 2 MoE layer expert parameters are replicated across all ranks by sharing a single tensor storage (with different offsets).
Since all MoE layer experts are stored in individual files for all ranks the default `torch.save()` behavior stores the underlying tensor's storage (even though a given expert is only covering a fraction of the storage).

Thus we end up with the following file structure per layer N:
```
layer_N_expert_E_mp_rank_00_model_states.pt
...
layer_N_expert_2_mp_rank_00_model_states.pt
layer_N_expert_1_mp_rank_00_model_states.pt
layer_N_expert_0_mp_rank_00_model_states.pt
```

Where _each_ file has the storage for all E experts across layer N, but the tensors in the files are only using a fraction of it.
When `torch.load()`ed each end up being tensors with independent storage.

It seems like using `.clone().detach()` when saving MoE expert tensors to individual files should address the issue.

But I was wondering about the reason behind making each expert and individual file for all ranks. E.g. if we stored a single file for expert layer for all ranks then the storage would have been reused for the loaded tensors and this single file would not be bloated.

I'll provide a test shortly, willing to contribute the fix.

**To Reproduce**

**Expected behavior**
Checkpoint files should be compact, no redundancy.

**ds_report output**
Please run `ds_report` to give us details about your setup.

**Screenshots**

**System info (please complete the following information):**
 - OS: Amazon Linux 4.14.252-195.483.amzn2.x86_64
 - GPU count and types: 8x A100 on a single machine
 - Interconnects (if applicable): single machine
 - Python version: 3.8
 - Any other relevant info about your setup: DeepSpeed 0.6.0, 0.7.4

**Launcher context**
DeepSpeed runner/launcher

**Docker context**
Custom docker image

**Additional context**
Using MoE layers, ZeRO 2

### PR DESCRIPTION
**Pull request by clumsy**
truncating expert param storage for checkpointing

### CODE REVIEW

File: tests/unit/simple_model.py [hunk 1]
```diff
@@ -43,19 +43,26 @@ class SimpleMoEModel(torch.nn.Module):
...
                            ep_size=ep_size,
                            use_residual=use_residual,
                            num_experts=num_experts,
                            k=1)
         self.cross_entropy_loss = torch.nn.CrossEntropyLoss()
 
     def forward(self, x, y):
-        hidden_dim = x
-        hidden_dim = self.linear(hidden_dim)
+        hidden_dim = self.linear(x)
```
Comment (tjruwase) [inline tests/unit/simple_model.py:63]: What is the rationale for this change?

**Comment by clumsy** [comment]
> 



> @clumsy please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> Contributor License Agreement

@microsoft-github-policy-service agree [company="Amazon"]

File: tests/unit/simple_model.py (hunk 1 above)
Comment (tjruwase) [reply tests/unit/simple_model.py:63]: Yes, you are correct. Thanks!

**Comment by mrwyattii** [comment]
> @tjruwase Could you please help me identify why some of these unit tests fail? It's mostly related to sparse attention which is seemingly unrelated to my change. And there are also 2 tests (but not all) failing with `CUDA out of memory` for MoE checkpointing which I didn't get for my device.

@clumsy We were seeing problems with our self-hosted CI runners yesterday. I believe they are resolved now. I have re-run the failing jobs, and they should pass this time.

**Comment by mrwyattii** [comment]
> Oh, that makes sense. Thanks, @mrwyattii! It looks like all but one check passed now. And the one that failed - looks like it failed the same way in a seemingly unrelated to my change place. Do we need to rerun it again?

Let me take a closer look at this today - not sure why it's failing

**Comment by mrwyattii** [comment]
Looks like tests are passing, with exception of one inference test that is getting OOM error. It's unrelated to any changes in this PR, so I'll merge it anyway
